File: controllers/ilqr/scripts/module/dynamics.py
#!/usr/bin/env python3

import rospy

import numpy as np
import casadi as ca

PI = 3.14

class Dynamics():
    def __init__(self, dt=.1, model="standard"):

        self.model = model
        self.dt = dt

        # State vector
        x = ca.SX.sym('x', 3)
            ## Standard
            # x_1 = x
            # x_2 = y
            # x_3 = theta 
            ## Tracking-error
            # e_1 = longitudinal error
            # e_2 = lateral error
            # e_3 = rotational error
        
        # Action control vector
        u = ca.SX.sym('u', 2) 
            # u_1 = linear velocity
            # u_2 = angular velocity

        # Dynamic function
        if (model == "standard"):

            x1_dot = u[0] * np.cos(x[2])
            x2_dot = u[0] * np.sin(x[2])
            x3_dot = u[1]
            
            x_dot = ca.veccat(x1_dot, x2_dot, x3_dot)

            f = x + x_dot * dt

        ## Tracking-error
        elif (model == "error-tracking"):

            e1_dot =  u[1] * x[1] - u[0]
            e2_dot = -u[1] * x[0]
            e3_dot = -u[1]

            e_dot = ca.veccat(e1_dot, e2_dot, e3_dot)

            f = x + e_dot * dt

        else:
            rospy.logerr("Not exist this dynamic model (%s)", model)

        # Dynamics derivatives
        Ak = ca.jacobian(f, x)
        Bk = ca.jacobian(f, u)

        # CasADi functions
        self._f = ca.Function('f', [x, u], [f])
        self._A = ca.Function('A_k', [x, u], [Ak])
        self._B = ca.Function('B_k', [x, u], [Bk])

    def f(self, x, u):
        return self._f(x, u)

    def f_prime(self, x, u):
        A_k = self._A(x, u)
        B_k = self._B(x, u)
        return A_k, B_k

File: controllers/ilqr/src/utils/dynamics.cpp
#include "ilqr/dynamics.h"

Dynamics::Dynamics()
{
    _dt = 0.1;
    _model = "";
}

Dynamics::Dynamics(double dt, std::string model)
{
    // Saving classes atributes
    _dt = dt;
    _model = model;

    // State space
    SX x = SX::sym("x", 3);
    SX u = SX::sym("u", 2);

    SX f;

    // Unicycle Kinematic Model
    if (model == "standard") 
    {
        SX x1_dot = u(0) * cos(x(2));
        SX x2_dot = u(0) * sin(x(2));
        SX x3_dot = u(1);
        
        SX x_dot = vertcat(x1_dot, x2_dot, x3_dot);

        f = x + x_dot * dt;
    }

    // Error-Tracking (Mobile point)
    else if (model == "error-tracking") 
    {
        SX e1_dot =  u(1) * x(1) - u(0);
        SX e2_dot = -u(1) * x(0);
        SX e3_dot = -u(1);

        SX e_dot = vertcat(e1_dot, e2_dot, e3_dot);

        f = x + e_dot * dt;
    }
    else {
        ROS_ERROR_STREAM("Not exist this dynamic model" << model);
    }

    // Dynamics derivatives (Aproxximation for A and B --> x_dot = Ax + Bu)
    SX f_x = jacobian(f, x);
    SX f_u = jacobian(f, u);

    // Dynamics Functions
    _f = Function("f", {x, u}, {f});
    _f_x = Function("f_x", {x, u}, {f_x});
    _f_u = Function("f_u", {x, u}, {f_u});
}

std::vector<DM> Dynamics::get_f(std::vector<DM> input) {
    return _f(input);
}

f_prime_t Dynamics::get_f_prime(std::vector<DM> input)
{
    _f_prime.f_x = Eigen::Matrix<double, 3, 3>::Map(DM::densify(_f_x(input).at(0)).nonzeros().data(), 3, 3);
    _f_prime.f_u = Eigen::Matrix<double, 3, 2>::Map(DM::densify(_f_u(input).at(0)).nonzeros().data(), 3, 2);

    return _f_prime;
}
